### PR TITLE
Add integration with Hashicorp Vault

### PR DESCRIPTION
**Pull request by felixgborrego**
Add support for secured environment variables stored in Vault.

Any environment var with a prefix `vault://key` is replaced by the secured value stored in the  [Vault](https://www.vaultproject.io) for the key before been passed as to Docker as an env var.  

Vault client gets its config from the environment `VAULT_ADDR` and `VAULT_TOKEN`

### CODE REVIEW

File: vault/vault.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if resp != nil && resp.StatusCode == 404 {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (relistan) [inline vault/vault.go]: The logic in here is hard to follow and looks like it's missing handlers for many possible HTTP status codes. Seems like it should be:
```
if err != nil {
    return nil, err
}
if resp == nil {
    return nil, err
}
defer resp.Body.Close()

if resp.StatusCode < 200 || resp.StatusCode > 299 {
    return nil, nil
}
```

File: vault/vault.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/hashicorp/vault/api"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+const VaultPathPrefix = "vault://"
+
+// Client to replace vault paths by the secret valued stored in the Hashicorp Vault
+type EnvVault interface {
```
Comment (relistan) [inline vault/vault.go]: Since we don't use this interface anywhere, this is normally not implemented as an interface and Impl. Just combine this and the Impl into an `EnvVault` struct with methods. We can easily turn it back into an interface later if needed (but probably won't).

File: vault/vault_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		Convey("isVaultPath() returns true for a vault key value", func() {
+			value := isVaultPath("vault://test")
+			So(value, ShouldBeTrue)
+
+			value = isVaultPath("vault://secret/key1")
+			So(value, ShouldBeTrue)
+		})
+
+		Convey("ReadSecretValue fails for an invalid key", func() {
+			_, err := envVault.ReadSecretValue("invalid path")
```
Comment (relistan) [inline vault/vault_test.go]: Do you want to validate that the first argument is nil here?

File: vault/vault_test.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+			So(err.Error(), ShouldEqual, "Invalid Vault path 'invalid path', expecting prefix 'vault://'")
+
+		})
+
+		Convey("ReadSecretValue fails for key not found", func() {
+			_, err := envVault.ReadSecretValue("vault://secure/notFoundKey")
+			So(err.Error(), ShouldEqual, "Path 'secure/notFoundKey' not found [VAULT_ADDR:http://test]")
```
Comment (relistan) [inline vault/vault_test.go]: It's not required, but I find that the tests are less brittle if you just compare for the part of the string you really need to match on vs the whole string. That way minor changes to the messages don't break the tests.

**Comment by relistan** [comment]
Solid, @felixgborrego! This is going to be a great addition. I made some minor comments above.

File: vault/vault.go (hunk 1 above)
Comment (relistan) [inline vault/vault.go]: Ideally we'd just look for "200" or whatever the client is supposed to return and barf on anything else.

File: vault/vault_test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+				StatusCode: 404,
+				Body:       ioutil.NopCloser(bytes.NewReader([]byte(""))),
+			},
+		}, nil
+	}
+
+	if strings.Contains(r.URL.Path, "secure/validKey") {
+		return &api.Response{
+			Response: &http.Response{
+				Body: ioutil.NopCloser(bytes.NewReader([]byte("{ \"data\":{ \"value\": \"secret_value\" }}"))),
```
Comment (mihaitodor) [inline vault/vault_test.go]: I'd use a [heredoc](https://godoc.org/github.com/MakeNowJust/heredoc) string here instead of escaping the double quotes.

**Comment by mihaitodor** [comment]
PS: you probably want to add `github.com/hashicorp/vault/api` to Godeps.

File: vault/vault.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return "", errors.Errorf("Unable to fetch '%s' [VAULT_ADDR:%s] error: %s", path, v.client.Address(), err.Error())
+	}
+
+	if secret == nil {
+		return "", errors.Errorf("Path '%s' not found [VAULT_ADDR:%s]", path, v.client.Address())
+	}
+
+	log.Debugf("Vault secret for path='%s', value: %s", path, secret.Data)
+	value, ok := secret.Data["value"].(string)
```
Comment (relistan) [inline vault/vault.go]: Pretty sure you won't be bale to access `value` outside the `if` block if you do that.

Comment (mihaitodor) [inline vault/vault.go]: ah, crap. You're right :) Ignore this suggestion :)

File: vault/vault_test.go (hunk 5 above)
Comment (relistan) [inline vault/vault_test.go]: Don't need a heredoc, just use backticks

Comment (mihaitodor) [inline vault/vault_test.go]: Yeah, that should suffice here :)

File: vault/vault.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+	if err != nil {
+		return "", errors.Errorf("Unable to fetch '%s' [VAULT_ADDR:%s] error: %s", path, v.client.Address(), err.Error())
+	}
+
+	if secret == nil {
+		return "", errors.Errorf("Path '%s' not found [VAULT_ADDR:%s]", path, v.client.Address())
+	}
+
+	log.Debugf("Vault secret for path='%s', value: %s", path, secret.Data)
```
Comment (felixgborrego) [inline vault/vault.go]: `secret.Data` is a ` map[string]interface{}` here. Just printing the full response in debug mode.

Comment (mihaitodor) [inline vault/vault.go]: Ah, in that case you most likely want to use "%+v" instead of "%s", which is intended only for strings.

**Comment by relistan** [comment]
:shipit:
